### PR TITLE
Reflection

### PR DESCRIPTION
**Pull request by schaumb**

### CODE REVIEW

File: src/chart/animator/styles.cpp [hunk 1]
```diff
@@ -74,41 +74,34 @@ void StyleMorphFactory::populate(::Anim::Group &group,
...
-	reinterpret_cast<Styles::Chart *>(pActual)->visit(*this);
+	Refl::visit(*this, *pSource, *pTarget, *pActual);
 }
 
 template <typename T>
-StyleMorphFactory &StyleMorphFactory::operator()(T &value,
-    const char *)
+auto StyleMorphFactory::operator()(
+    const T &source, const T &target, const T &value
+) -> std::void_t<decltype(interpolate(source, target, value))>
```
Comment (schaumb) [inline src/chart/animator/styles.cpp:83]: `interpolate(*source, *target, 0.0)`

File: src/base/refl/auto_enum.h [hunk 2]
```diff
@@ -25,30 +25,10 @@ error_str(std::string_view name, std::string_view code)
...
-	    func.find_last_not_of("abcdefghijklmnopqrstuvwxyz"
-	                          "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
-	                          "0123456789_");
-	constexpr std::string_view res = func.substr(val + 1);
-	return res.length() > 0 && (res[0] < '0' || res[0] > '9')
-	         ? res
-	         : std::string_view{};
-}
-
+namespace Detail {
```
Comment (schaumb) [inline src/base/refl/auto_enum.h:28]: Apply clang-format

File: src/base/style/paramregistry.h [hunk 3]
```diff
@@ -18,12 +18,29 @@ namespace Style
...
+		    fromString(
+		        [t](Root &r, const std::string &str) mutable
+		        {
+			        auto &e = t(r);
+			        e = Conv::parse<std::remove_cvref_t<decltype(e)>>(
+			            str);
+		        })
+		{}
+
+		Accessor(const Accessor &) = delete;
```
Comment (schaumb) [inline src/base/style/paramregistry.h:39]: Add deleted  move constructors too

File: src/base/style/paramregistry.h [hunk 4]
```diff
@@ -62,81 +57,51 @@ template <typename Root> class ParamRegistry
...
-		{
-			auto &value = *reinterpret_cast<T *>(
-			    reinterpret_cast<std::byte *>(&root) + this->offset);
-
-			if (s == "null")
-				value.reset();
-			else
-				value = Conv::parse<typename T::value_type>(s);
+		if (path.empty()) {
+			return std::pair{accessors.begin(), accessors.end()};
```
Comment (schaumb) [inline src/base/style/paramregistry.h:71]: Make return value to ranges::views::subrange

File: src/base/style/paramregistry.h [hunk 5]
```diff
@@ -62,81 +57,51 @@ template <typename Root> class ParamRegistry
...
 		}
-	};
+		else {
+			return std::pair{accessors.lower_bound(path + "."),
```
Comment (schaumb) [inline src/base/style/paramregistry.h:74]: Create prefix string comparator class, use equal_range here

File: src/chart/animator/styles.cpp [hunk 6]
```diff
@@ -55,16 +55,16 @@ StyleMorphFactory::StyleMorphFactory(const Styles::Chart &source,
 {
 	dry = true;
 	needed = false;
-	pSource = reinterpret_cast<const std::byte *>(&source);
-	pTarget = reinterpret_cast<const std::byte *>(&target);
-	pActual = reinterpret_cast<std::byte *>(&actual);
+	pSource = &source;
```
Comment (schaumb) [inline src/chart/animator/styles.cpp:58]: Move it to the initialized list

**Review by schaumb — COMMENTED** [review]
-

File: src/base/refl/auto_struct.h [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+#include <utility>
+#include <limits>
+#include <functional>
+
+#include "auto_name.h"
+
+namespace Refl
+{
+
+template <class... Ts> struct tuple
```
Comment (schaumb) [inline src/base/refl/auto_struct.h:16]: It is necessary?

File: src/base/refl/auto_struct.h [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+	template<class T, class U>
+	constexpr auto operator<<(const T& op, U&& arg)
+		-> decltype(op(std::forward<U&&>(arg)))
+	 {
+		return op(std::forward<U&&>(arg));
+	}
+}
+
+template<std::size_t I, class T>
+struct NthBase : T {
```
Comment (schaumb) [inline src/base/refl/auto_struct.h:295]: Not needed

File: src/base/refl/auto_struct.h [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+	Wrapper(T) -> Wrapper<T>;
+
+	template<auto A>
+	using IC = std::integral_constant<decltype(A), A>;
+
+    template <class T, class Visitor>
+    constexpr auto get_member_names()
+    {
+        return []<std::size_t ...Ix>(std::index_sequence<Ix...>) {
+			constexpr auto M = getAllFilteredMembers<T, Visitor>();
```
Comment (schaumb) [inline src/base/refl/auto_struct.h:556]: format.

File: src/chart/animator/styles.cpp [hunk 10]
```diff
@@ -1,70 +1,33 @@
...
-{
-	dry = true;
-	needed = false;
-	pSource = reinterpret_cast<const std::byte *>(&source);
-	pTarget = reinterpret_cast<const std::byte *>(&target);
-	pActual = reinterpret_cast<std::byte *>(&actual);
-}
+	: dry{true}
+	, needed{false}
+    , pActual(std::addressof(actual))
```
Comment (schaumb) [inline src/chart/animator/styles.cpp:13]: format

**Comment by schaumb** [comment]
The parent main commit wasm size:
```
"wasm_make": 346172 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm
"wasm_make": 151578 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm.gz
```

This PR wasm size:
```
"wasm_make": 359297 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm
"wasm_make": 155554 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm.gz
```

**Review by schaumb — COMMENTED** [review]
❌ no ::members() - necessary for fontParent (Font class). Others are removed.

- [x] ".tpp" for ParamReg cons
- [x] fontParent visit - or remove


- [x] Check ParamMerger, StyleMorphFactory are necessary 
  - When I merge it to the paramregistry, the size is increasing. So it is good separately.
- [x] CompositeRec move to compositeOP

- [x] std::tuple_size_v<Loophole::aggregate_types_t<T>> > 0 always true
- [x] StyleMorphFactory::isNeeded make const -> psource == ptarget
  - equality operator checks faulty all member, which is not intended.

❌ Font::Style interpolation by oblique - created another ticket

**Review by simzer — DISMISSED** [review]

**Comment by simzer** [comment]
What is the final binary size increase with this PR?

**Comment by schaumb** [comment]
342718 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm
150456 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm.gz

After merge:

355204 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm
153944 : /workspace/build/cmake-wasm/weblib/cvizzu.wasm.gz

(+13 kb)

**Review by simzer — APPROVED** [review]

**Pull request merged**